### PR TITLE
New C++ Api: Refactor check macros.

### PR DESCRIPTION
**Pull request by aniemetz**
This introduces more CHECK macros for ease of use.

### CODE REVIEW

**Review by ajreynol — CHANGES_REQUESTED** [review]
One general comment I have is whether it is overkill to define macros for some things that will probably not be used more than once. For example, CVC4_API_TERM_CHECK_TERMS_WITH_SORTS_COMPARABLE_TO.  This makes the code a bit more tedious to follow.  Is there a reason to make a macro for this?

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -5274,43 +6198,33 @@ Term Solver::defineFunRec(const std::string& symbol,
...
   {
-    type = nm->mkFunctionType(domain_types, type);
+    domain_sorts.push_back(bv.getSort());
   }
-  Node fun = d_nodeMgr->mkVar(symbol, type);
-  std::vector<Node> ebound_vars = Term::termVectorToNodes(bound_vars);
-  d_smtEngine->defineFunctionRec(fun, ebound_vars, *term.d_node, global);
-  return Term(this, fun);
-  CVC4_API_SOLVER_TRY_CATCH_END;
+  Sort s = domain_sorts.empty() ? sort : mkFunctionSort(domain_sorts, sort);
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:6216]: Indeed! Good catch!

**Comment by aniemetz** [comment]
> One general comment I have is whether it is overkill to define macros for some things that will probably not be used more than once. For example, CVC4_API_TERM_CHECK_TERMS_WITH_SORTS_COMPARABLE_TO. This makes the code a bit more tedious to follow. Is there a reason to make a macro for this?

Yes. The idea here was to make it easier to separate checks and the remainder of the code. Some of them we might even reuse in the future. Alternatively, we could use macros for the checks that occur more often, but then we would have to iterate over arguments again for additional checks, e.g., for this one https://github.com/CVC4/CVC4/pull/6123/files#diff-902f7604d0650139a4d3a195f77fa90e784779503655cd18cc96a21832cd668cR1120.

**Comment by aniemetz** [comment]
Maybe adding yet another iteration is no issue at all, though. The number of arguments in vectors or sets will be rather low, typically. This will only make a difference for millions of calls to these API functions.

**Comment by ajreynol** [comment]
Yeah, I can see your point about separating the macros.  On a related note, would it make sense to move the macro definitions to another file? (Or is this not possible?)

**Comment by aniemetz** [comment]
We can move them, yes. Do you want this on this PR?
And, do you want me to reduce the number of checks, at the cost of more loop iterations for checks?

**Comment by aniemetz** [comment]
Also, would this all be easier if I split the PR by Sort, Term, Solver?

**Comment by aniemetz** [comment]
Let me try to split this into steps.

**Pull request closed without merging**